group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,399 @@
+                    unassignedPartitions.remove(partition);
+                }
+
+                // Only re-add to the end of the queue if it's still in the unfilledMembers map
+                if (unfilledMembers.containsKey(memberId)) {
+                    roundRobinMembers.add(memberId);
+                }
+            }
+        }
+    }
+
+    /**
+     * Allocates the unassigned partitions to available members.
+     *
+     * If the rack-aware strategy is enabled, partitions are attempted to be assigned back to their current owners first.
+     *
+     * If a partition couldn't be assigned to its current owner due to quotas or
+     * if the rack-aware strategy is not enabled, the partitions are allocated to members in a round-robin fashion.</p>
+     */
+    private void unassignedPartitionsRoundRobinAssignment() {
+        Queue<String> roundRobinMembers = new LinkedList<>(unfilledMembers.keySet());
+        Iterator<TopicIdPartition> partitionIterator = unassignedPartitions.iterator();
+
+        while (partitionIterator.hasNext()) {
+            TopicIdPartition partition = partitionIterator.next();
+            boolean assigned = false;
+
+            if (useRackAwareStrategy && currentPartitionOwners.containsKey(partition)) {
+                String prevOwner = currentPartitionOwners.get(partition);
+                if (unfilledMembers.containsKey(prevOwner)) {
+                    assignPartitionToMember(prevOwner, partition);
+                    assigned = true;
+                    partitionIterator.remove();
+                    if (!unfilledMembers.containsKey(prevOwner)) {
+                        roundRobinMembers.remove(prevOwner);
+                    }
+                }
+            }
+
+            // Only re-add the member to the end of the queue if it's still available for assignment.
+            for (int i = 0; i < unfilledMembers.size() && !assigned; i++) {
+                String memberId = roundRobinMembers.poll();
+                if (unfilledMembers.get(memberId) > 0) {
+                    assignPartitionToMember(memberId, partition);
+                    assigned = true;
+                    partitionIterator.remove();
+                }
+                if (unfilledMembers.containsKey(memberId)) {
+                    roundRobinMembers.add(memberId);
+                }
+            }
+        }
+    }
+
+    /**
+     * Assigns the specified partition to the given member.
+     *
+     * <p>
+     * If the member has met their allocation quota, the member is removed from the
+     * tracking map of members with their remaining allocations.
+     * Otherwise, the count of remaining partitions that can be assigned to the member is updated.
+     * </p>
+     *
+     * @param memberId      The Id of the member to which the partition will be assigned.
+     * @param partition     The partition to be assigned.
+     */
+    private void assignPartitionToMember(String memberId, TopicIdPartition partition) {
```

Comment:
It is in the javadoc

Reply:
* If the member has met their allocation quota, the member is removed from the
     * tracking map of members with their remaining allocations.
     * Otherwise, the count of remaining partitions that can be assigned to the member is updated.